source/dofs/dof_renumbering.cc

```diff
@@ -1191,50 +1191,65 @@ namespace DoFRenumbering
+            current_dof_index
+              = compute_hierarchical_recursive<dim> (current_dof_index,
+                                                     new_indices,
+                                                     cell->child (c),
+                                                     locally_owned_dof_indices);
         }
       else
         {
           if (cell->is_locally_owned())
             {
+              // first get the existing DoF indices
               const unsigned int dofs_per_cell = cell->get_fe().dofs_per_cell;
               std::vector<types::global_dof_index> local_dof_indices (dofs_per_cell);
               cell->get_dof_indices (local_dof_indices);
 
+              // then loop over the existing DoF indices on this cell
+              // and see whether it has already been re-numbered (it
+              // may have been on a face or vertex to a neighboring
+              // cell that we have encountered before). if not,
+              // give it a new number and store that number in the
+              // output array (new_indices)
+              //
+              // if this is a parallel triangulation and a DoF index is
+              // not locally owned, then don't touch it. since
+              // we don't actually *change* DoF indices (just record new
+              // numbers in an array), we don't need to worry about
+              // the decision whether a DoF is locally owned or not changing
+              // as we progress in renumbering DoFs -- all adjacent cells
+              // will always agree that a DoF is locally owned or not
               for (unsigned int i = 0; i < dofs_per_cell; ++i)
-                {
-                  if (locally_owned.is_element (local_dof_indices[i]))
-                    {
-                      // this is a locally owned DoF, assign new number if not assigned a number yet
-                      unsigned int idx = locally_owned.index_within_set (local_dof_indices[i]);
-                      if (new_indices[idx] == numbers::invalid_dof_index)
-                        {
-                          new_indices[idx] = locally_owned.nth_index_in_set (next_free);
-                          next_free++;
-                        }
-                    }
-                }
+                if (locally_owned_dof_indices.is_element (local_dof_indices[i]))
+                  {
+                    // this is a locally owned DoF, assign new number if not assigned a number yet
+                    const unsigned int idx = locally_owned_dof_indices.index_within_set (local_dof_indices[i]);
+                    if (new_indices[idx] == numbers::invalid_dof_index)
+                      {
+                        new_indices[idx] = locally_owned_dof_indices.nth_index_in_set (current_dof_index);
+                        ++current_dof_index;
+                      }
+                  }
             }
         }
-      return next_free;
+
+      return current_dof_index;
```

Comment:
except that this rename makes it less clear that what we return is the next free dof index. Maybe leave a comment at the beginning of the function that states this?

Reply:
I renamed the variable further, and also left a comment to this effect.